ql/src/java/org/apache/hadoop/hive/ql/io/SkippingTextInputFormat.java

```diff
@@ -0,0 +1,136 @@
+  }
+
+  @Override
+  protected FileSplit makeSplit(Path file, long start, long length, String[] hosts) {
+    long cachedStart = getCachedStartIndex(file);
+    long cachedEnd = getCachedEndIndex(file);
+    if (cachedStart > start + length) {
+      return new NullRowsInputFormat.DummyInputSplit(file);
+    } else if (cachedStart > start) {
+      length = length - (cachedStart - start);
+      start = cachedStart;
+    }
+    if (cachedEnd > -1 && cachedEnd < start) {
+      return new NullRowsInputFormat.DummyInputSplit(file);
+    } else if (cachedEnd < start + length) {
+      length = cachedEnd - start;
+    }
+    return super.makeSplit(file, start, length, hosts);
+  }
+
+  @Override
+  protected FileSplit makeSplit(Path file, long start, long length, String[] hosts, String[] inMemoryHosts) {
+    long cachedStart = getCachedStartIndex(file);
+    long cachedEnd = getCachedEndIndex(file);
+    if (cachedStart > start + length) {
+      return new NullRowsInputFormat.DummyInputSplit(file);
+    } else if (cachedStart > start) {
+      length = length - (cachedStart - start);
+      start = cachedStart;
+    }
+    if (cachedEnd > - 1 && cachedEnd < start) {
+      return new NullRowsInputFormat.DummyInputSplit(file);
+    } else if (cachedEnd < start + length) {
+      length = cachedEnd - start;
+    }
+    return super.makeSplit(file, start, length, hosts, inMemoryHosts);
+  }
+
+  private long getCachedStartIndex(Path path) {
+    Long startIndexForFile = startIndexMap.get(path);
+    if (startIndexForFile == null) {
+      try {
+        FSDataInputStream fis = path.getFileSystem(conf).open(path);
+        for (int j = 0; j < headerCount; j++) {
+          fis.readLine();
+          // back 1 byte because readers skip the entire first row if split start is not 0
+          startIndexForFile = fis.getPos() - 1;
+        }
+      } catch (IOException e) {
+        startIndexForFile = 0L;
+      }
+      startIndexMap.put(path, startIndexForFile);
+    }
+    return startIndexForFile;
+  }
+
+  private long getCachedEndIndex(Path path) {
+    Long endIndexForFile = endIndexMap.get(path);
+    if (endIndexForFile == null) {
+      try {
+        final long bufferSectionSize = 1024;
+        long bufferSectionEnd = path.getFileSystem(conf).getFileStatus(path).getLen();
+        long bufferSectionStart = Math.max(0, bufferSectionEnd - bufferSectionSize);
+        Queue<Long> lineEndBuffer = new ArrayDeque<Long>(headerCount + 1);
+        FSDataInputStream fis = path.getFileSystem(conf).open(path);
+        fis.seek(bufferSectionStart);
+        while (bufferSectionEnd > bufferSectionStart) {
+          fis.seek(bufferSectionStart);
+          long pos = fis.getPos();
+          while (pos < bufferSectionEnd) {
+            fis.readLine();
+            pos = fis.getPos();
+            if (pos <= bufferSectionEnd) {
+              if (lineEndBuffer.size() > footerCount) {
+                lineEndBuffer.poll();
```

Comment:
I think, there is a problem in this logic. The queue is expected to have positions added in descending order. But, let's say, if first batch doesn't make the queue full, then 2nd batch would  break this descending order rule. For instance, it can be 15->20->0->5->10 where 10 being the head. Now, in this case, instead of 20 being the endIndex we pick 10 which is incorrect.